timers/timers.go

```diff
@@ -0,0 +1,255 @@
+// Package timers provides a timer facility similar to the standard library's.
+// It's designed to be faster, but less accurate.
+package timers
+
+import (
+	"sync"
+	"time"
+	"unsafe"
+
+	"github.com/andres-erbsen/clock"
+	"github.com/uber-go/atomic"
+)
+
+const cacheline = 64
+
+const (
+	// State machine for timers.
+	stateScheduled = iota
+	stateExpired
+	stateCanceled
+)
+
+// A Timer is a handle to a deferred operation.
+type Timer struct {
+	f        func()
+	deadline uint64 // in ticks
+	state    atomic.Int32
+	next     *Timer
+	tail     *Timer // only set on head
+}
+
+// Stop cancels the deferred operation. It returns whether or not the
+// cancellation succeeded.
+func (t *Timer) Stop() bool {
+	if t == nil {
+		return true // canceling a nil timer always succeeds
+	}
+	return t.state.CAS(stateScheduled, stateCanceled)
+}
+
+func (t *Timer) expire() bool {
+	return t.state.CAS(stateScheduled, stateExpired)
+}
+
+func (t *Timer) pushOne(head *Timer) *Timer {
+	head.next, head.tail = nil, nil
+	return t.push(head)
+}
+
+func (t *Timer) push(head *Timer) *Timer {
+	// link lists
+	if head == nil {
+		return t
+	}
+	if head.tail == nil {
+		head.next = t
+	} else {
+		head.tail.next = t
+	}
+
+	// set tail
+	if t == nil {
+		return head
+	}
+	if t.tail == nil {
+		head.tail = t
+	} else {
+		head.tail = t.tail
+		t.tail = nil
+	}
+	return head
+}
+
+type bucket struct {
+	sync.Mutex
+	timers *Timer
+	// Avoid false sharing:
+	// http://mechanical-sympathy.blogspot.com/2011/07/false-sharing.html
+	_ [cacheline - unsafe.Sizeof(sync.Mutex{}) - unsafe.Sizeof(&Timer{})]byte
+}
+
+type bucketList struct {
+	mask    uint64
+	buckets []bucket
+}
+
+func (bs *bucketList) Schedule(deadline uint64, f func()) *Timer {
+	if f == nil {
```

Comment:
Do we need to support this use case? The standard library does not, I don't think we should either,
https://play.golang.org/p/OdxBnyCzXQ

Seems like a code bug if someone is scheduling timers to do nothing.